modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C

```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
+  Real denom_const = -n[3][2]*(n[5][0] - n[5][1]) - n[3][1]*(-n[5][0] + n[5][2]) + n[3][0]*(-n[5][1] + n[5][2]);
+  Real denom_coeff = -n[3][2]*(n[5][0] - n[5][1]) - n[3][1]*(n[5][0] + n[5][2]) + n[3][0]*(n[5][1] + n[5][2]);
+  Real denom = denom_const + denom_coeff*sinphi;
+  Real residual = -x + numer/denom;
+
+  Real deriv_phi;
+  Real deriv_coh;
+  Real jacobian;
+  Real tol = std::pow(_f_tol/mag_E/10.0, 2);
+  unsigned int iter = 0;
+  do {
+    do {
+      deriv_phi = dphi(intnl_old + dpm[3]);
+      deriv_coh = dcohesion(intnl_old + dpm[3]);
+      jacobian = -1 + (numer_coeff1*deriv_coh*cosphi - numer_coeff1*coh*sinphi*deriv_phi + numer_coeff2*cosphi*deriv_phi)/denom - numer*denom_coeff*cosphi*deriv_phi/denom/denom;
+      x -= residual/jacobian;
+      if (iter > _max_iters) // not converging
+      {
+        nr_converged = false;
+        return false;
+      }
+
+      sinphi = std::sin(phi(intnl_old + x));
+      cosphi = std::cos(phi(intnl_old + x));
+      coh = cohesion(intnl_old + x);
+      cohcos = coh*cosphi;
+      numer = numer_const + numer_coeff1*cohcos + numer_coeff2*sinphi;
+      denom = denom_const + denom_coeff*sinphi;
+      residual = -x + numer/denom;
+      iter ++;
+    } while (residual*residual > tol);
+
+    // now must ensure that yf[3] and yf[5] are both "zero"
+    Real dpm3minusdpm5 = (2*(eigvals[0] - eigvals[1]) + x*(n[3][1] - n[3][0] + n[5][1] - n[5][0]))/(n[3][0] - n[3][1] + n[5][1] - n[5][0]);
+    dpm[3] = (x + dpm3minusdpm5)/2.0;
+    dpm[5] = (x - dpm3minusdpm5)/2.0;
+
+    for (unsigned i = 0 ; i < 3 ; ++i)
+      returned_stress(i, i) = eigvals[i] - dpm[3]*n[3][i] - dpm[5]*n[5][i];
+    yieldFunctionEigvals(returned_stress(0, 0), returned_stress(1, 1), returned_stress(2, 2), sinphi, cohcos, yf);
+  } while (yf[3]*yf[3] > _f_tol*_f_tol && yf[5]*yf[5] > _f_tol*_f_tol);
+
+
+  // so the NR process converged, but we must
+  // check Kuhn-Tucker
+  nr_converged = true;
+
+  if (dpm[3] < -ep_plastic_tolerance || dpm[5] < -ep_plastic_tolerance)
+    // Kuhn-Tucker failure.    This is a potential weak-point: if the user has set _f_tol quite large, and ep_plastic_tolerance quite small, the above NR process will quickly converge, but the solution may be wrong and violate Kuhn-Tucker.
+    return false;
+
+  if (yf[0] > _f_tol || yf[1] > _f_tol || yf[2] > _f_tol || yf[4] > _f_tol)
+    // Kuhn-Tucker failure
+    return false;
+
+  // success
+  dpm[0] = dpm[1] = dpm[2] = dpm[4] = 0;
+  return true;
 }
 
-Real
-TensorMechanicsPlasticMohrCoulombMulti::psi(const Real internal_param) const
+
+bool
+TensorMechanicsPlasticMohrCoulombMulti::returnEdge010100(const std::vector<Real> & eigvals, const std::vector<std::vector<Real> > & n, std::vector<Real> & dpm, RankTwoTensor & returned_stress, const Real & intnl_old, Real & sinphi, Real & cohcos, const Real & initial_guess, const Real & mag_E, bool & nr_converged, const Real & ep_plastic_tolerance, std::vector<Real> & yf) const
```

Comment:
Generalise returnEdge000101 and returnEdge010100 to just "returnEdge".  I actually did this for returnTip110100 and returnTip010101 (and made the returnTip function) which is only a little obscure because of that generalisation.  I did also write a "returnEdge" function but it contained so many "if" statements that it was just too obscure - it's already a nightmare with those stupid Mathematica-generated expressions!  So i'll consider this, but am not that hopeful.